rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+/// !REVIEW: I've tried to closely follow SymCrypts API calls and semantics making some simplifications, but I think we have a chance to simplify further if we choose to.
+/// My questions / Suggestions:
+/// 1. We've introduced the idea of the RsaKeyPair / RsaPublicKey state, which can hold a lot of data, and makes a lot of the calls to the underlying SymCrypt code redundant. most of the accessors no longer
+/// needed to call SymCrypt since the same data is stored on the state or is assumed
+/// 2. to keep this thread safe, I've combined the alloc / generate and alloc / set pattern to just one function
```

Comment:
>to keep this thread safe, I've combined the alloc / generate and alloc / set pattern to just one function [](http://example.com/codeflow?start=7&length=105)

I don't think this makes the object fully thread safe. For instance, if one thread were to call `set_key_pair` at the same time as another thread were calling one of the export functions, bad things could happen.

Reply:
The set_key_pair creates the RSA object. I believe Khang is going to rename.